File: SICP/Lecture_6_full/Code/Java/Stream/src/fayelab/sicp/stream/java8stream/Procedures.java
package fayelab.sicp.stream.java8stream;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.IntStream;
import java.util.stream.Stream;

import static java.util.Arrays.asList;
import static java.util.stream.Collectors.toList;

public class Procedures
{
    @SuppressWarnings("unchecked")
    static <T> Stream<T> enumTree(Object obj)
    {
        if(obj instanceof BiTuple)
        {
            BiTuple<?, ?> biTuple = (BiTuple<?, ?>)obj;
            return Stream.concat(enumTree(biTuple.getElement1()), enumTree(biTuple.getElement2()));
        }
        
        return Stream.of((T)obj);
    }

    static Stream<Integer> enumInterval(int low, int high)
    {
        return IntStream.rangeClosed(low, high).boxed();
    }
    
    public static int sumOddsSquare(BiTuple<?, ?> biTree)
    {
        Stream<Integer> s = enumTree(biTree);
        return s.filter(x -> isOdd(x))
                .map(x -> square(x))
                .reduce((x, y) -> x + y)
                .get();
    }
    
    public static List<Integer> oddFibs(int n)
    {
        List<Integer> result = new ArrayList<>();
        
        enumInterval(0, n).map(idx -> asList(idx, fib(idx)))
                          .filter(idxAndFib -> isOdd(idxAndFib.get(1)))
                          .collect(() -> result, 
                                   (acc, idxAndFib) -> acc.add(idxAndFib.get(0)),
                                   (acc1, acc2) -> acc1.addAll(acc2));

        return result;
    }
    
    public static List<List<Integer>> primeSumPairs(int n)
    {
        return enumInterval(2, n).flatMap(i -> enumInterval(1, i - 1).map(j -> asList(i, j))
                                                                     .filter(pair -> isPrime(pair.get(0) + pair.get(1))))
                                 .collect(toList());
    }
    
    public static List<List<Integer>> triples(int n)
    {
        return enumInterval(3, n).flatMap(i -> enumInterval(2, i - 1).flatMap(j -> enumInterval(1, j - 1).map(k -> asList(i, j, k))))
                                 .collect(toList());
    }
    
    private static boolean isOdd(int n)
    {
        return n % 2 == 1;
    }
    
    private static int square(int n)
    {
        return n * n;
    }
    
    private static int fib(int n)
    {
        int a = 0;
        int b = 1;
        int s = 0;
        
        if(n == 0)
        {
            return a;
        }
        
        if(n == 1)
        {
            return b;
        }
        
        for(int i = 0; i <= n - 2; i++)
        {
            s = a + b;
            a = b;
            b = s;
        }
        
        return s;
    }
    
    private static boolean isPrime(int n)
    {
        return enumInterval(2, n - 1).noneMatch(i -> n % i == 0);
    }

    public static <T1, T2> BiTuple<T1, T2> biTuple(T1 element1, T2 element2)
    {
        return new BiTuple<>(element1, element2);
    }

    static class BiTuple<T1, T2>
    {
        private T1 element1;
        private T2 element2;
        
        public BiTuple(T1 element1, T2 element2)
        {
            this.element1 = element1;
            this.element2 = element2;
        }
    
        public T1 getElement1()
        {
            return element1;
        }
    
        public T2 getElement2()
        {
            return element2;
        }
    }
}

File: SICP/Lecture_6_full/Code/Python/bif/procedures.py
## Procedures with built-in functions
def sum_odds_square(biTree):
    return reduce(lambda x, y: x + y, 
                  map(lambda x: square(x), filter(lambda x: isodd(x), enum_tree(biTree))), 
                  0)

def odd_fibs(n):
    return [idx for idx in enum_interval(0, n) if isodd(fib(idx))]

def prime_sum_pairs(n):
    return [(i, j) for i in enum_interval(2, n) for j in enum_interval(1, i - 1) if isprime(i + j)]

def triples(n):
    return [(i, j, k) for i in enum_interval(3, n) for j in enum_interval(2, i - 1) for k in enum_interval(1, j - 1)]

def enum_tree(obj):
    result = []

    if isinstance(obj, tuple):
        return enum_tree(obj[0]) + enum_tree(obj[1])

    result.append(obj)
    return result

def enum_interval(low, high):
    return range(low, high + 1)

## Auxiliary functions
def isodd(n):
    return n % 2 != 0

def square(n):
    return n * n

def fib(n):
    if n == 0:
        return 0
    if n == 1:
        return 1

    return fib(n - 1) + fib(n - 2)

def isprime(n):
    return all(map(lambda x: n % x != 0, range(2, n)))

## Tests
def test_enum_tree():
    assert [1, 2, 7, 19, 12, 14] == enum_tree(((1, (2, 7)), (19, (12, 14))))
    assert [1, 2, 3, 7, 19, 12, 13, 14] == enum_tree(((1, ((2, 3), 7)), (19, (12, (13, 14)))))
    print 'test_enum_tree ok'

def test_enum_interval():
    assert [2, 3, 4, 5] == enum_interval(2, 5)
    assert [] == enum_interval(3, 1)
    print 'test_enumInterval ok'

def test_sum_odds_square():
    assert 411 == sum_odds_square(((1, (2, 7)), (19, (12, 14))))
    print 'test_sum_odds_square ok'

def test_odd_fibs():
    assert [1, 2, 4, 5] == odd_fibs(6)
    print 'test_odd_fibs ok' 

def test_prime_sum_pairs():
    assert [(2, 1), (3, 2), (4, 1), (4, 3), (5, 2)] == prime_sum_pairs(5)
    print 'test_prime_sum_pairs ok'

def test_triples():
    assert [(3, 2, 1),
            (4, 2, 1), (4, 3, 1), (4, 3, 2),
            (5, 2, 1), (5, 3, 1), (5, 3, 2), (5, 4, 1), (5, 4, 2), (5, 4, 3)] == triples(5)
    print 'test_triples ok'

def test():
    test_enum_tree()
    test_enum_interval()
    test_sum_odds_square()
    test_odd_fibs()
    test_prime_sum_pairs()
    test_triples()

    print 'test ok'

if __name__ == '__main__':
    test()
